### PR TITLE
confidential asset types

### PR DESCRIPTION
**Pull request by garbageslam**
This PR implements the transaction math part of confidential token ids.
MCIP (still draft) is here: https://github.com/mobilecoinfoundation/mcips/pull/25

Still TODO:
- [x] Add the proof of knowledge of opening
- [x] Rebase on master
- [x] Use the mob id constant and token id types in master
- [ ] Rename Amount -> MaskedAmount (that may come in a separate PR afterwards)
- [x] More tests

---

More TODO
- [x] Remove proof of knowledge of opening
- [x] Undo breaking changes to protos
- [x] Use the block version feature test in transaction builder and validation
- [x] Make sample paykit support multiple asset balances
- [x] Make mobilecoind track token ids
- [x] Make slam and fog distro respect token ids
- [x] rebase again
- [x] Make it possible to bootstrap with token ids, and test in CD

### CODE REVIEW

File: api/proto/external.proto
Comment (jcape) [inline api/proto/external.proto:230]: Is this a breaking change?

File: transaction/core/src/amount/proof_of_opening.rs
Comment (jcape) [inline transaction/core/src/amount/proof_of_opening.rs:21]: Reading through the slides, is this better described as a proof of openability?

**Comment by jcape** [comment]
As a general comment, the the `AmountData {}` struct seems really awkward to use, maybe just wrap it up entirely with `Amount`?

Like this *should* be represented as:
```
pub struct Amount<T: Token> {
    value: u64,
    _type: PhantomData<T>,
}
```

And just use the token ID when serializing/deserializing. I realize this is more complex than that, but I worry that we're opening the door to a lot of type confusion bugs by not letting the compiler enforce this kind of thing wherever possible.

**Comment by garbageslam** [comment]
> I dunno, I feel like this should be represented as:

One of the design requirements is, if we ship this functionality, it should be possible to stand up new token ids in prod without signing a new enclave.

If we make AmountData a template over the token id as you describe, then there is no way the enclave can create fee outputs for tokens that didn't exist when it was signed, because those types were not part of the build at that time.. So this proposal is incompatible with the requirements. Let me know if you still have questions about this.

**Comment by garbageslam** [comment]
> And just use the token ID when serializing/deserializing. I realize our uses are more complex than that, but I worry that we're opening the door to a lot of type confusion bugs by not letting the compiler enforce this kind of thing wherever possible.

I think actually, the best mechanism we have for catching those bugs is the proof of opening, which is checked both by the signer after they construct the Tx, and by the consensus verifier. Even if there is a bug or a weak API in the TransactionBuilder where tokens of different types get mixed up, build the transaction will fail because they will be unable to create a valid proof of opening.

File: transaction/core/src/amount/proof_of_opening.rs
Comment (garbageslam) [reply transaction/core/src/amount/proof_of_opening.rs:21]: Yeah I'm open to other names for this thing

I think "proof of opening" jibes with other things I've read?

We could also call it like "proof of generator" or "proof of pedersen base" or something like this? That's sort of more to the point, it proves that the prover built this commitment over a particular set of pedersen generators?

File: api/proto/external.proto
Comment (garbageslam) [reply api/proto/external.proto:236]: Also intentional

File: api/proto/external.proto
Comment (jcape) [reply api/proto/external.proto:230]: That doesn't require changing the types or tags of existing fields.

Comment (garbageslam) [reply api/proto/external.proto:230]: I think the organization is much better if the fee is next to the token id, it makes it easier to read and understand. Is there a problem with that?

**Comment by eranrund** [comment]
On the topic of breaking changes, I am wondering what the plan is for getting this deployed. If we deploy this to the network before SDKs upgrade, then they will no longer be able to submit transactions.
If we want to update SDKs first, they will need to support constructing both old and new txs and decide based on the node they connect to whether to construct v1 or v2.

It would be nice if we could somehow make the `Tx`/`TxPrefix` change non-breaking so that for MOB transactions stuff just continues to work. Not sure how feasible that is with the signature change. But if we are making a breaking change I do think we should plan for either:
1) Having Consensus support accepting both old and new tx types for a time period (which might allow us to deploy this change prior to any SDK updates?)
2) Having the SDKs support both current and new Tx construction, and figure out how they decide which version to submit.

**Comment by garbageslam** [comment]
I don't know right now how we can make this not breaking. If we allow people to omit proof of opening, then they could mix token ids. So once we start having multiple token ids, the old transaction types are not valid.

We could maybe like, duplicate the transaction-core crate and have a `transaction-core v2` crate. But it would be very painful.

I think it would be simpler to make the clients figure out based on the MRENCLAVE they get during attestation whether to use the v1 transaction builder or the v2 transaction builder

**Comment by garbageslam** [comment]
We could try to make the enclave depend on two different revisions of transaction core, that might avoid actually copy pasting all the code. But it's still going to be very complicated.

**Comment by eranrund** [comment]
That does mean that we need to figure out how to expose two versions of the TransactionBuilder for the SDKs.

**Comment by garbageslam** [comment]
@eranrund they might be able to just link an old and new version of libmobilecoin?

**Comment by garbageslam** [comment]
maybe there is a better way, maybe we can accept the old Txs as long as all the inputs and outputs are missing the masked_token_id, and then skip the proof of opening.

**Comment by garbageslam** [comment]
and then in a future revision, we require the masked token id and proof of opening?

that might be a better direction for this PR, gonna think on it some more

**Comment by eranrund** [comment]
> @eranrund they might be able to just link an old and new version of libmobilecoin?

I would be surprised, due to symbol name collision. Might be able to do some horrible linker tricks to get around it but I am not excited about that...

I also have no idea how the Java JNI stuff would behave.

File: transaction/core/src/amount/mod.rs
Comment (mfaulk) [inline transaction/core/src/amount/mod.rs:89]: This comment looks copy-pasted. Should be something like "The token_id is XORed...", no?

File: transaction/core/src/amount/proof_of_opening.rs
Comment (mfaulk) [reply transaction/core/src/amount/proof_of_opening.rs:23]: The best description I have so far is "a proof of knowledge of a Pedersen commitment opening". That terminology is used here, https://eprint.iacr.org/2020/061.pdf

So, ProofOfOpening seems ok to me.

File: transaction/core/src/encrypted_fog_hint.rs
Comment (mfaulk) [inline transaction/core/src/encrypted_fog_hint.rs:33]: Length

File: transaction/core/src/lib.rs
Comment (mfaulk) [inline transaction/core/src/lib.rs:7]: Thank you

File: transaction/core/src/ring_signature/mlsag.rs
Comment (mfaulk) [inline transaction/core/src/ring_signature/mlsag.rs:469]: Nice

File: transaction/core/src/ring_signature/mod.rs
Comment (mfaulk) [inline transaction/core/src/ring_signature/mod.rs:54]: I haven't been able to find a reference for "orthogonal" base points. I pulled up some definitions of Pedersen commitments and found a couple of good descriptions:

"Given the two generators G and H, no one knows the discrete log for second generator H with respect to first generator G, meaning no one knows a y such that y ∗ G = H"

or 

Computational soundness (binding): if the generators g, h are independently generated, then finding a discrete logarithm relation between them is as hard as breaking the Discrete Log Problem.

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (mfaulk) [inline transaction/core/src/ring_signature/rct_bulletproofs.rs:417]: Thanks. This looks much better.

File: transaction/core/src/amount/mod.rs
Comment (garbageslam) [reply transaction/core/src/amount/mod.rs:89]: yes thank you

File: api/proto/external.proto
Comment (garbageslam) [reply api/proto/external.proto:230]: we're going to redesign per discussion so that this isn't a breaking change, and to the extent there is, navigate that via the block version bumping

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (garbageslam) [reply transaction/core/src/ring_signature/rct_bulletproofs.rs:418]: I'm now not sure if I should actually do this, since it's a breaking change and if we remove proof of opening this may not have to be a breaking change...

pondering it

Comment (garbageslam) [reply transaction/core/src/ring_signature/rct_bulletproofs.rs:418]: I guess that since we are doing the block version thing and it's a hard cut-over anyways we might as well keep this

File: transaction/core/src/amount/mod.rs
Comment (garbageslam) [inline transaction/core/src/amount/mod.rs:144]: @voloshyn i noticed that in a lot of the SDKs, this function is basically being reimplemented from scratch using `get_value_mask` etc, i think it will be more maintainable if we try to use this version instead

File: libmobilecoin/src/transaction.rs
Comment (eranrund) [inline libmobilecoin/src/transaction.rs:96]: This feels very fishy to me... is this PR the right place to change the behavior? 
Is this used by the SDKs? Maybe its better to remove it and have the SDK deal with it when they uprev?

File: mobilecoind/src/service.rs
Comment (eranrund) [inline mobilecoind/src/service.rs:567]: What should we do with it?

File: slam/src/main.rs
Comment (eranrund) [inline slam/src/main.rs:504]: Shouldn't it get it from the SpendableTxOuts? and panic if they are mixing multiple token ids?

File: slam/src/main.rs
Comment (eranrund) [inline slam/src/main.rs:344]: Maybe a good idea to add to the token id to SpendableTxOut

File: slam/src/main.rs
Comment (garbageslam) [reply slam/src/main.rs:504]: yeah that seems better

File: mobilecoind/src/service.rs
Comment (garbageslam) [reply mobilecoind/src/service.rs:567]: I guess the `UnspentTxOut` object should store it

Comment (garbageslam) [reply mobilecoind/src/service.rs:567]: we could store an `AmountData` instead of a value to try to keep the token id with the value

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (eranrund) [inline transaction/core/src/ring_signature/rct_bulletproofs.rs:620]:
```suggestion
            let mut params = SignatureParams::random(block_version, num_inputs, num_mixins, &mut rng);
```

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (eranrund) [inline transaction/core/src/ring_signature/rct_bulletproofs.rs:642]:
```suggestion
            let mut params = SignatureParams::random(block_version, num_inputs, num_mixins, &mut rng);
```

**Review by eranrund — COMMENTED** [review]
I went over this, and it all seems good to me with the exception of a relatively long list of TODOs that we need to keep track of in the SDKs/mobilecoind/sample-paykit/etc. Are you keeping track of it in Asana or a GH issue that summarizes all the followup work needed?

Has the most recent version of this been deployed and performance looked at? There have been concerns about not caching the generators if I recall correctly.

Thank you!

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (garbageslam) [reply transaction/core/src/ring_signature/rct_bulletproofs.rs:620]: i wonder why cargo fmt isn't doing this

Comment (eranrund) [reply transaction/core/src/ring_signature/rct_bulletproofs.rs:620]: I think because it's inside a macro call

File: libmobilecoin/src/transaction.rs
Comment (garbageslam) [reply libmobilecoin/src/transaction.rs:96]: we chatted about this with the SDK team, the current code here is a bug. they may have fixed it and perhaps i need to rebase. will follow up with them

Comment (garbageslam) [reply libmobilecoin/src/transaction.rs:96]: @mobilecoinfoundation/sdk ^lmk what you think about this

**Comment by garbageslam** [comment]
> Are you keeping track of it in Asana or a GH issue that summarizes all the followup work needed?

I'm mostly updating the PR description with new TODO items. I need to make a list of things to happen after this in GH i guess

File: transaction/core/src/ring_signature/rct_bulletproofs.rs
Comment (garbageslam) [inline transaction/core/src/ring_signature/rct_bulletproofs.rs:390]: @mfaulk @isis-mc lmk how you guys feel about this, Matt I think you already commented positively on switching to Merlin. To keep compat though we have to use the old version until block version = 3 lands.

File: libmobilecoin/src/transaction.rs
Comment (garbageslam) [reply libmobilecoin/src/transaction.rs:96]: according to Adam this function is never actually called

File: transaction/core/src/range_proofs/mod.rs
Comment (garbageslam) [inline transaction/core/src/range_proofs/mod.rs:8]: @mfaulk this is where i added note about the generator playing the role of the proof of opening, we could expand on this I guess

**Comment by garbageslam** [comment]
Screenshot at 2022-03-02 16-01-12

**Comment by garbageslam** [comment]
Screenshot at 2022-03-02 16-01-38

**Comment by garbageslam** [comment]
tx/s looks about the same,
tx completion time is a bit worse, but this is also with a slam that has twice as many transactions, so that may be a reason

File: fog/distribution/src/main.rs
Comment (eranrund) [inline fog/distribution/src/main.rs:157]: Nit: This should not require block_version to be mut - unwrap_or_else can be used to fallback to the latest block version.

File: fog/distribution/tests/bootstrap.rs
Comment (eranrund) [inline fog/distribution/tests/bootstrap.rs:15]: This should use https://doc.rust-lang.org/std/fs/fn.remove_dir_all.html

File: fog/distribution/tests/bootstrap.rs
Comment (eranrund) [inline fog/distribution/tests/bootstrap.rs:9]: This is a pretty funky test. Is it generating all the keys and stuff in the target directory?
It should create a temporary directory that gets cleaned up after the test is done, and not rely on deleting stuff that might happen to exist in the current directory.

File: util/generate-sample-ledger/tests/bootstrap.rs
Comment (eranrund) [inline util/generate-sample-ledger/tests/bootstrap.rs:1]: Similar to the other test, this should not have to delete stuff that might've been left over from a previous run. It also doesn't do any validation that the binary actually produced anything useful, so personally I am not finding much value here.

**Review by eranrund — COMMENTED** [review]
I went over the recent changes and made some comments. Did not spot any major blockers.
I am ready to hit approve once there are no more changes - but I would also like more than one other approval for this particular PR since it touches very critical paths. I think three reviewers who are familiar with the nature of this change would be appropriate.

Thank you for taking the time to document stuff and improve the utilities, thats super nice and will make our lives better!

File: fog/distribution/tests/bootstrap.rs
Comment (garbageslam) [reply fog/distribution/tests/bootstrap.rs:9]: Ok we can do that

File: util/generate-sample-ledger/tests/bootstrap.rs
Comment (garbageslam) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: It is needed to have some kind of integration test in the util generate sample keys crate, otherwise cargo test will not build the binary, and then the other test will fail

Comment (garbageslam) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: It's testing that the binary runs without crashing -- we could do more validation, any obvious thing worth testing?

File: mobilecoind/src/sync.rs
Comment (garbageslam) [reply mobilecoind/src/sync.rs:387]: I was going to try to do it by filtering at each call like that, but maybe the api is too large to do it that way sanely. I'm just not sure what happens if we filter when adding things to database, then they restart the program with a different token id. Probably they won't compute the right balance

Comment (garbageslam) [reply mobilecoind/src/sync.rs:387]: They would have to delete the ledger and resync I guess

File: mobilecoind/src/service.rs
Comment (garbageslam) [reply mobilecoind/src/service.rs:334]: They could restart the program with a different token id I think

File: fog/distribution/tests/bootstrap.rs
Comment (garbageslam) [reply fog/distribution/tests/bootstrap.rs:9]: I added this test to try to reduce my iteration times in CD

File: util/generate-sample-ledger/tests/bootstrap.rs
Comment (garbageslam) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: I guess we could also merge these crates into one

Comment (eranrund) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: Relying on having a unit test just so that the binary gets built and can be used by a unit test in another crate is exactly what I would call funky...
If we want to do this, the right approach is to move the functionality into a library, and call that through a `dev-dependencies` dependency.

File: mobilecoind/src/sync.rs
Comment (eranrund) [reply mobilecoind/src/sync.rs:387]: Good point on restarting with a different token id. Okay, with that in mind I think it does make sense to to store all the matched txouts, and have the API filter out to the configured token id (until we add proper support for multiple token ids from a single mobilecoind).

File: fog/distribution/tests/bootstrap.rs
Comment (eranrund) [reply fog/distribution/tests/bootstrap.rs:9]: At the very least it is worth explaining the purpose of this.

File: util/generate-sample-ledger/tests/bootstrap.rs
Comment (jcape) [inline util/generate-sample-ledger/tests/bootstrap.rs:15]: Same comment re: future introduction of `remove_dir_all` for these.

**Review by jcape — COMMENTED** [review]
Mostly LGTM, there are a few minor things noted in the comments.

File: transaction/core/src/range_proofs/mod.rs
Comment (garbageslam) [reply transaction/core/src/range_proofs/mod.rs:32]: i think the formatter reflowed my comments, will try to figure out how to get it not to do that

File: api/proto/external.proto
Comment (garbageslam) [reply api/proto/external.proto:133]: i can probably make some kind of test for that, i think its possible to enumerate over the protobuf enum

File: fog/sample-paykit/src/cached_tx_data/mod.rs
Comment (garbageslam) [reply fog/sample-paykit/src/cached_tx_data/mod.rs:861]: i guess i'm anticipating that Eran wants to rename `AmountData -> Amount` and `Amount -> MaskedAmount`

File: mobilecoind/api/proto/mobilecoind_api.proto
Comment (garbageslam) [reply mobilecoind/api/proto/mobilecoind_api.proto:130]: i don't think so, i think the field #10 is not actually serialized

File: util/generate-sample-ledger/tests/bootstrap.rs
Comment (garbageslam) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: i wish that cargo would give me a way to depend on binaries from other crates :(

i really prefer testing the actual binary if i can because the whole point of this is to try to catch issues before waiting for a CD deploy cycle

Comment (garbageslam) [reply util/generate-sample-ledger/tests/bootstrap.rs:1]: maybe this just shouldn't be a cargo test if cargo can't accomodate this, maybe should build the binaries and then just run a script

File: mobilecoind/api/proto/mobilecoind_api.proto
Comment (jcape) [reply mobilecoind/api/proto/mobilecoind_api.proto:130]: Yeah, my concern is about the jump from tag 7 to tag 10... it feels like there were tags 7-9 at some point in the past, and they were removed but not documented.

Comment (garbageslam) [reply mobilecoind/api/proto/mobilecoind_api.proto:130]: i see, so maybe this should be #11 ? I can look in git history.

thanks

File: libmobilecoin/src/transaction.rs
Comment (garbageslam) [reply libmobilecoin/src/transaction.rs:200]: So, I agree, but I don't maintain the libmobilecoin API, and usually when i make changes that impact libmobilecoin, my goal is, do not change the libmobilecoin header and do not break their build. Instead just leave comments that would help Adam adjust it when he gets to it. I can make a github ticket for him.

**Review by isis-mc — APPROVED** [review]
I didn't look at the changes to the test suite code super closely, but the changes to the transaction maths look fine to me.

File: transaction/core/src/amount/commitment.rs
Comment (garbageslam) [reply transaction/core/src/amount/commitment.rs:2]: We are just importing this name from the Bulletproofs crate that you worked on:

```
pub use bulletproofs_og::{BulletproofGens, PedersenGens};
```

I'm inclined to just use the upstream name to avoid confusion

Comment (garbageslam) [reply transaction/core/src/amount/commitment.rs:2]: https://github.com/dalek-cryptography/bulletproofs/blob/1a10ce1a5b87299014658770346b376a858e7691/src/generators.rs#L30

**Comment by garbageslam** [comment]
This has passed deploy (build network #46), I think we should merge it if there are no outstanding issues.

I didn't get the bootstrap with custom token ids to work, that seems to break deployment and I'm not sure why. But I think we can test that functionality end to end once minting is merged.

File: libmobilecoin/src/transaction.rs
Comment (garbageslam) [reply libmobilecoin/src/transaction.rs:96]: @the-real-adammork feel free to request that I do something different, I'm just trying to avoid doing something that would break your build, and making sure that there are tickets for follow up work

Comment (eranrund) [reply libmobilecoin/src/transaction.rs:96]: If its not called can we remove it? Having incorrect code is worse than having a breaking build in my opinion.

File: transaction/core/src/token.rs
Comment (eranrund) [inline transaction/core/src/token.rs:48]: Nice

**Review by eranrund — APPROVED** [review]
LGTM, thank you for all your hard work on this.
I would prefer if this one gets more than two approvals before merging, preferably two more approvals from people intimately familiar with the codebase.

**Review by mfaulk — APPROVED** [review]

**Comment by mfaulk** [comment]
Thank you for all of your work on this, Chris!

**Review by sugargoat — APPROVED** [review]
This is awesome - thanks for the great work and prototype. Will make sure the MCIP is up to date with this as well!

**Review by mfaulk — APPROVED** [review]

**Pull request merged**
